### PR TITLE
thermal assistant icon position based on 30s avg

### PR DESCRIPTION
**Pull request by GeorgeDowning20**
Adjust the position of the glider icon in the thermal assistant according to the vario average. makes the thermal assistant more intuitive to use.

### CODE REVIEW

**Comment by DanD222** [comment]
@GeorgeDowning20 I’m having a hard time imagining what the difference is, would you have a screenshot by any chance please?

**Comment by GeorgeDowning20** [comment]
> @GeorgeDowning20 I’m having a hard time imagining what the difference is, would you have a screenshot by any chance please?
Screenshot from 2023-09-19 23-59-56
Screenshot from 2023-09-19 23-59-12

The position of the glider icon is the 30s average:

1) Where the thermal assistant trace is normalized to fit in the circle, this upgrade helps the user quantify the magnitude of the trace relative to their average. useful for gauging the significance of spikes in the trace. 

2) The assistant is more intuitive: 
-if the lift is stronger than your average (best bit), the glider is shown to be inside the trace, and highlighted blue.
-if the lift is less strong than your trace (worst bit), the glider is shown to be outside of the trace.

This makes it significantly clearer to identify which way you need move to get better centered.

**Comment by lordfolken** [comment]
I like the change very much.
Two things
  Lets document it somewhere 
  A ring would be indeed very nice

**Comment by GeorgeDowning20** [comment]
> Lets document it somewhere

as in the manual?

**Comment by lordfolken** [comment]
Ideally both. And the help text in the enable disable feature.
But best would be intuitive recognition. 
Potential idea: Maybe a 30s label beside the plane icon?

**Comment by GeorgeDowning20** [comment]
> plane
Screenshot from 2023-09-20 18-45-21

not sure if I like the Macready circle. What do you think? It could be useful but clutters things a bit.
I have also added a smoothing function.

**Comment by lordfolken** [comment]
Instead of MC, how about you hook the circle to the plane icon and adapt its size to the 30s average?

Otherwise the circle would have to have a different color, or change the color whenever the climb value exceeds the average.

**Comment by DanD222** [comment]
> 
> not sure if I like the Macready circle. What do you think? It could be useful but clutters things a bit.

I quite like the idea, I think it would look a lot better if the colour of the MC circle was different though, and in a contrasting colour to the thermal shape colour.

**Comment by GeorgeDowning20** [comment]
> > not sure if I like the Macready circle. What do you think? It could be useful but clutters things a bit.
> 
> I quite like the idea, I think it would look a lot better if the colour of the MC circle was different though, and in a contrasting colour to the thermal shape colour.

i have had a play with this... also not a huge fan. can't find a colour scheme that doesn't clash. my plan is to change the color of the plane icon

**Comment by MaxKellermann** [comment]
I don't get it. I have no idea what the change is and how it is supposed to improve something. I don't understand the screenshots.

And in my opinion, the thermal assistant certainly should not have this MC ring. Ideally, the MC should be the same as the average lift, and then both shapes overlap in confusing ways, adding clutter, but without any practical value.

**Comment by GeorgeDowning20** [comment]
> I don't get it. I have no idea what the change is and how it is supposed to improve something. I don't understand the screenshots.

commit 1:
-Apply a smoothing filter to the thermal assistant trace. This improves the readability, where it can be difficult in weak or -gusty thermals.

Commit 2:
- The glider is positioned on the graph according to the average climb rate. This is a change from the original where the glider is positioned at the "max lift" mark, which is a useless measure. This improves the feature because the thermal assistant becomes more intuitive to use.

> And in my opinion, the thermal assistant certainly should not have this MC ring. Ideally, the MC should be the same as the average lift, and then both shapes overlap in confusing ways, adding clutter, but without any practical value.

MC ring idea dropped. in theory, this is useful for observing the best part of the lift. If your average climb rate starts to drop, you can interpolate if it will increase again in the next turn or if you just leave. However, I can't find a non-cluttered way of implementing this.

**Comment by MikeBravo15** [comment]
How about to have only MC ring?

**Comment by MaxKellermann** [comment]
>     * Apply a smoothing filter to the thermal assistant trace. This improves the readability, where it can be difficult in weak or gusty thermals.

Sure, smoothing makes things look nicer, but also increases latency a lot and discards a lot of detail. I don't think this is a good idea. Showing all the gory details (including all the gusts), is the whole point of the thermal assistant.

>     * The glider is positioned on the graph according to the average climb rate. This is a change from the original where the glider is positioned at the "max lift" mark, which is a useless measure. This improves the feature because the thermal assistant becomes more intuitive to use.

I have no idea what that means. How can a glider be positioned according to a climb rate?

> MC ring idea dropped. in theory, this is useful for observing the best part of the lift. If your average climb rate starts to drop, you can interpolate if it will increase again in the next turn or if you just leave. However, I can't find a non-cluttered way of implementing this.

Maybe choose the line color or fill color depending on whether that part of the line is above or below the MC (or this thermal's average). Make it more similar to the trace.

**Comment by DanD222** [comment]
> How can a glider be positioned according to a climb rate?

Here's my understanding of the situation - the red glider icons are where the glider would be positioned, given the glider's 30 second average climb was 0 kts and 7.8 kts.

The green circle is my understanding of the MC circle, which I don't think is all that bad - it shows me where the thermal is stronger than MC; the question is whether the additional information of `thermal > MC` in addition to `thermal > now` is worth the slight additional degradation in clarity.

image


> Showing all the gory details (including all the gusts), is the whole point of the thermal assistant.

My first thought regarding this is whether gusts that happened in the last circle would be useful/relevant to coring the thermal - I would think that only the best _consistent_ part of the climb is ultimately of interest in centering a thermal, so in practice I foresee the need to mentally "unimagine" them to get the best results in centering.

**Comment by GeorgeDowning20** [comment]
> My first thought regarding this is whether gusts that happened in the last circle would be useful/relevant to coring the thermal - I would think that only the best _consistent_ part of the climb is ultimately of interest in centering a thermal, so in practice I foresee the need to mentally "unimagine" them to get the best results in centering.

if the sampling rate was much higher, seeing gust could be really useful. You would be able to see the turbulence signature around cores. In the mean time, gust are not useful to centre cores.

**Comment by GeorgeDowning20** [comment]
> Sure, smoothing makes things look nicer, but also increases latency a lot and discards a lot of detail. I don't think this is a good idea. Showing all the gory details (including all the gusts), is the whole point of the thermal assistant.

the smoothing filter has been applied, around the unit circle, Y = 0.5z + 0.25z+1 + 0.25z-1. Therefore introduces no latency, just smooths out gusts.

if you don’t like I will remove.

The trouble with the existing systems is: although the positions on the graph for gusts is technically correct, guests are represented by too much surface area. This masks where the less strong but consistent lift is.

**Comment by DanD222** [comment]
> if the sampling rate was much higher

This is determined by the GPS position update rate, if I understand correctly? Probably tending to over-complicate things again, but I like the idea of the smoothing to be switchable in preferences - one could try if one prefers the thermal assistant with or without smoothing, and could make use of the gust information in case of a high update rate position source.

File: src/Gauge/ThermalAssistantRenderer.cpp [hunk 1]
```diff
@@ -88,12 +88,33 @@ ThermalAssistantRenderer::NormalizeLift(double lift, double max_lift)
...
+  canvas.DrawText(p, text);
+}
+
+static void
+DrawCircleLabelVSpeed(Canvas &canvas, PixelPoint p, double value) noexcept
+{
+  TCHAR buffer[10];
+  FormatUserVerticalSpeed(value, buffer, ARRAY_SIZE(buffer));
+  DrawCircleLabel(canvas, p, buffer);
+}
```
Comment (MaxKellermann) [inline src/Gauge/ThermalAssistantRenderer.cpp:108]: Creating duplicate code is not acceptable. Don't do that.

**Comment by MaxKellermann** [comment]
> Here's my understanding of the situation - the red glider icons are where the glider would be positioned, given the glider's 30 second average climb was 0 kts and 7.8 kts. As the glider's 30 second average changes, I would expect the glider icon to move along the x axis (please correct me if I'm wrong).

I don't expect it to do that. This feature does not make sense to me at all. I don't get why the glider should move anywhere. The only reason why it's there is to show the pilot in one quick glance whether he's on the left edge of the thermal (i.e. circling right) or on the right edge (circling left).

Moving the glider symbol around makes it harder to see that with one quick look. Not only that, this position doesn't even encode any useful information.

Also I don't get why this commit diff is so large. All it should do, according to the description, is change the X position, but there is so much more code. What does it do? Is it worth taking the time to try to read and understand it? I don't have time to waste, and I wasted so much time on reading PRs which the author would never finish, so I'm cautious where I spend the little time I have.

**Comment by GeorgeDowning20** [comment]
> > Here's my understanding of the situation - the red glider icons are where the glider would be positioned, given the glider's 30 second average climb was 0 kts and 7.8 kts. As the glider's 30 second average changes, I would expect the glider icon to move along the x axis (please correct me if I'm wrong).
> 
> I don't expect it to do that. This feature does not make sense to me at all. I don't get why the glider should move anywhere. The only reason why it's there is to show the pilot in one quick glance whether he's on the left edge of the thermal (i.e. circling right) or on the right edge (circling left).
> 
> Moving the glider symbol around makes it harder to see that with one quick look. Not only that, this position doesn't even encode any useful information.
> 
> Also I don't get why this commit diff is so large. All it should do, according to the description, is change the X position, but there is so much more code. What does it do? Is it worth taking the time to try to read and understand it? I don't have time to waste, and I wasted so much time on reading PRs which the author would never finish, so I'm cautious where I spend the little time I have.

Something weird is going on with the gitt diff... i have moved a function in-order to access an undeclared function. it should look like this...

Screenshot from 2023-09-22 20-23-51

I will declare the existing undeclared functions. and move the edited function back. the code should be really simple

File: src/Gauge/ThermalAssistantRenderer.cpp [hunk 2]
```diff
@@ -17,6 +17,13 @@
 
 #include <algorithm> // for std::clamp()
 
+static void
+DrawCircleLabel(Canvas &canvas, PixelPoint p,
+                tstring_view text) noexcept;
+
+static void
+DrawCircleLabelVSpeed(Canvas &canvas, PixelPoint p, double value) noexcept;
```
Comment (MaxKellermann) [inline src/Gauge/ThermalAssistantRenderer.cpp:25]: Undeclared? What does that mean?

**Comment by GeorgeDowning20** [comment]
fixed up the git diff, so is now more presentable. Sorry about that.

> I don't expect it to do that. This feature does not make sense to me at all. I don't get why the glider should move anywhere. The only reason why it's there is to show the pilot in one quick glance whether he's on the left edge of the thermal (i.e. circling right) or on the right edge (circling left).


The reason is: that the thermal assistant is effectively a graph. The icon is placed on the graph and therefore represents a value. At present this value is the 'max lift in the last turn'. This is a useless bit of information. By encoding the 30s average as this value, produces a handful of advantages as described above.

File: src/Gauge/ThermalAssistantRenderer.cpp [hunk 3]
```diff
@@ -113,6 +115,14 @@ ThermalAssistantRenderer::PaintRadarPlane(Canvas &canvas) const
...
                   p.At(-Layout::FastScale(small ? 2 : 4),
                        +Layout::FastScale(small ? 2 : 4)));
+
+  if (small)
+    return;
+
+  if (!circling.TurningLeft()) 
+    DrawCircleLabelVSpeed(canvas, p + PixelSize{-45, 0}, vario.average);
+  else 
+    DrawCircleLabelVSpeed(canvas, p + PixelSize{+45, 0}, vario.average);
```
Comment (MaxKellermann) [inline src/Gauge/ThermalAssistantRenderer.cpp:125]: Why is this here? This is not described anywhere, your commit message doesn't mention it.

File: src/Gauge/ThermalAssistantRenderer.cpp [hunk 4]
```diff
@@ -96,11 +96,13 @@ ThermalAssistantRenderer::NormalizeLift(double lift, double max_lift)
...
 void
-ThermalAssistantRenderer::PaintRadarPlane(Canvas &canvas) const
+ThermalAssistantRenderer::PaintRadarPlane(Canvas &canvas, double max_lift) const
 {
+  int normalised_average = NormalizeLift(vario.average, max_lift) * radius;
+
   canvas.Select(look.plane_pen);
 
-  PixelPoint p = mid.At(circling.TurningLeft() ? (int)radius : (int)-radius,
+  PixelPoint p = mid.At(circling.TurningLeft() ? (int)normalised_average : (int)-normalised_average,
```
Comment (MaxKellermann) [inline src/Gauge/ThermalAssistantRenderer.cpp:105]: What's the point of casting an `int` to an `int`?

File: src/Gauge/ThermalAssistantRenderer.cpp (hunk 3 above)
Comment (GeorgeDowning20) [reply src/Gauge/ThermalAssistantRenderer.cpp:125]: lordfolken suggested it? i will refactor PR to include

**Comment by DanD222** [comment]
> Not only that, this position doesn't even encode any useful information.

Here's my interpretation of the additional information:
image

File: src/Gauge/ThermalAssistantRenderer.cpp (hunk 2 above)
Comment (GeorgeDowning20) [reply src/Gauge/ThermalAssistantRenderer.cpp:25]: perhaps my terminology is wrong. Initially the 'PaintRadarPlane' radar function did not have access to these functions. My initial approach was to move the function but this caused diff problems. Therefore I have declared the function at the top. please can you suggest a better method?

**Comment by GeorgeDowning20** [comment]
> > Not only that, this position doesn't even encode any useful information.
> 
> Here's my interpretation of the additional information: image

I have removed the MC ring as Max strongly objects "And in my opinion, the thermal assistant certainly should not have this MC ring."

File: src/Gauge/ThermalAssistantRenderer.cpp (hunk 2 above)
Comment (MaxKellermann) [reply src/Gauge/ThermalAssistantRenderer.cpp:25]: You can move functions if they need to be moved. Just make it a separate commit (and let the commit message explain why this move is needed), or else you confuse reviewers.
Adding unnecessary forward declarations just clutters the code.

Comment (GeorgeDowning20) [reply src/Gauge/ThermalAssistantRenderer.cpp]: roger. thanks for your help. sorted

File: src/Gauge/ThermalAssistantRenderer.cpp [hunk 5]
```diff
@@ -127,6 +127,14 @@ ThermalAssistantRenderer::PaintRadarPlane(Canvas &canvas, double max_lift) const
...
                   p.At(-Layout::FastScale(small ? 2 : 4),
                        +Layout::FastScale(small ? 2 : 4)));
+
+  if (small)
+    return;
+
+  if (!circling.TurningLeft()) 
+    DrawCircleLabelVSpeed(canvas, p + PixelSize{-45, 0}, vario.average);
+  else 
+    DrawCircleLabelVSpeed(canvas, p + PixelSize{+45, 0}, vario.average);
```
Comment (MaxKellermann) [inline src/Gauge/ThermalAssistantRenderer.cpp:137]: These two lines are duplicate, the only difference is the sign. This should be parametrized. Also where does this `45` come from? Don't hard-code pixel sizes!

**Comment by MaxKellermann** [comment]
I've cherry-picked the first two commits, even though I don't like the feature. The third commit needs to be fixed, but I still don't quite get the point, and I don't like it either; it just adds more graphical clutter for something that is supposed to be very simple.

**Comment by GeorgeDowning20** [comment]
> I've cherry-picked the first two commits, even though I don't like the feature. The third commit needs to be fixed, but I still don't quite get the point, and I don't like it either; it just adds more graphical clutter for something that is supposed to be very simple.

Thank you for your help getting this up to scratch. The first two commits represent my original idea.

The label and MC ring, are additional stuff that has come out of this thread. I will continue working on them if wanted else i will focus my attention elsewhere.

**Comment by DanD222** [comment]
> I have removed the MC ring as Max strongly objects "And in my opinion, the thermal assistant certainly should not have this MC ring." but thank you, this shows the additional information well.
> 
> The label and MC ring, are additional stuff that has come out of this thread. I will continue working on them if wanted else i will focus my attention elsewhere.

Fair enough, let it serve "for illustration purposes only" then. I still see some value in the additional information gained, and I like the idea of possible better coring with the smoothed thermal graph on low-update screens, but Better flarm colours are significantly more important than either of those.

There's also potentionally low-hanging fruit here? https://github.com/XCSoar/XCSoar/issues/1206
